File: VB/WindowsApplication1/MyNavBarControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports DevExpress.XtraNavBar

Namespace WindowsApplication1
	Public Class MyNavBarControl
		Inherits NavBarControl

		Private Function NextGroup(ByVal group As NavBarGroup, ByVal forward As Boolean) As NavBarGroup
			If Groups.Count = 0 Then '|| (Groups.Count == 1 && group != null)
				Return Nothing
			End If
			If group Is Nothing Then
				If (forward) Then
					Return Groups(0)
				Else
					Return Groups(Groups.Count - 1)
				End If
			End If

			If forward Then
				If Groups.IndexOf(group) < Groups.Count - 1 Then
					Return Groups(Groups.IndexOf(group) + 1)
				Else
					Return Groups(0)
				End If
			ElseIf Groups.IndexOf(group) > 0 Then
				Return Groups(Groups.IndexOf(group) - 1)
			Else
				Return Groups(Groups.Count - 1)
			End If
		End Function

		Private Function FindNearItemLink(ByVal forward As Boolean) As NavBarItemLink
			Dim link As NavBarItemLink = SelectedLink
			Dim group As NavBarGroup
			If link Is Nothing Then
				group = NextGroup(Nothing, forward)
			Else
				group = link.Group
			End If
			If group Is Nothing Then
				Return Nothing
			End If
			Dim startGroup As NavBarGroup = group
			Dim i As Integer
			If link IsNot Nothing Then
				i = group.ItemLinks.IndexOf(link)
			Else
				If (forward) Then
					i = -1
				Else
					i = group.ItemLinks.Count
				End If
			End If
			Dim repeat As Boolean = False
			Do
				If forward Then
					If i < group.ItemLinks.Count - 1 Then
						Return group.ItemLinks(i + 1)
					Else
						group = NextGroup(group, forward)
						repeat = i > 0
						i = -1
					End If
				Else
					If i > 0 AndAlso group.ItemLinks.Count > 0 Then
						Return group.ItemLinks(i - 1)
					Else
						group = NextGroup(group, forward)
						repeat = i < group.ItemLinks.Count - 1
						i = group.ItemLinks.Count
					End If
				End If
			Loop While group IsNot startGroup OrElse repeat
			Return Nothing
		End Function

		Public Sub SelectNextLink(ByVal forward As Boolean)
			SelectedLink = FindNearItemLink(forward)
			If SelectedLink IsNot Nothing AndAlso (Not SelectedLink.Group.Expanded) Then
				SelectedLink.Group.Expanded = True
			End If
			ViewInfo.MakeLinkVisible(SelectedLink)
		End Sub

		Protected Overrides Function ProcessDialogKey(ByVal keyData As Keys) As Boolean
			Dim key As Keys = keyData And ((Not Keys.Modifiers))
			Select Case key
				Case Keys.Down
					SelectNextLink(True)
				Case Keys.Up
					SelectNextLink(False)
				Case Else
					Return MyBase.ProcessDialogKey(keyData)
			End Select
			Return False
		End Function
	End Class
End Namespace

File: CS/WindowsApplication1/MyNavBarControl.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using DevExpress.XtraNavBar;

namespace WindowsApplication1
{
    public class MyNavBarControl : NavBarControl
    {

        private NavBarGroup NextGroup(NavBarGroup group, bool forward)
        {
            if (Groups.Count == 0) //|| (Groups.Count == 1 && group != null)
                return null;
            if (group == null)
                return (forward) ? Groups[0] : Groups[Groups.Count - 1];

            if (forward)
            {
                if (Groups.IndexOf(group) < Groups.Count - 1)
                    return Groups[Groups.IndexOf(group) + 1];
                else
                    return Groups[0];
            }
            else if (Groups.IndexOf(group) > 0)
                return Groups[Groups.IndexOf(group) - 1];
            else
                return Groups[Groups.Count - 1];
        }

        private NavBarItemLink FindNearItemLink(bool forward)
        {
            NavBarItemLink link = SelectedLink;
            NavBarGroup group;
            if (link == null)
                group = NextGroup(null, forward);
            else
                group = link.Group;
            if (group == null)
                return null;
            NavBarGroup startGroup = group;
            int i;
            if (link != null)
                i = group.ItemLinks.IndexOf(link);
            else
                i = (forward) ? -1 : group.ItemLinks.Count;
            bool repeat = false;
            do
            {
                if (forward)
                {
                    if (i < group.ItemLinks.Count - 1)
                        return group.ItemLinks[i + 1];
                    else
                    {
                        group = NextGroup(group, forward);
                        repeat = i > 0;
                        i = -1;
                    }
                }
                else
                {
                    if (i > 0 && group.ItemLinks.Count > 0)
                        return group.ItemLinks[i - 1];
                    else
                    {
                        group = NextGroup(group, forward);
                        repeat = i < group.ItemLinks.Count - 1;
                        i = group.ItemLinks.Count;
                    }
                }
            }
            while (group != startGroup || repeat);
            return null;
        }

        public void SelectNextLink(bool forward)
        {
            SelectedLink = FindNearItemLink(forward);
            if (SelectedLink != null && !SelectedLink.Group.Expanded)
                SelectedLink.Group.Expanded = true;
            ViewInfo.MakeLinkVisible(SelectedLink);
        }

        protected override bool ProcessDialogKey(Keys keyData)
        {
            Keys key = keyData & (~Keys.Modifiers);
            switch (key)
            {
                case Keys.Down:
                    SelectNextLink(true);
                    break;
                case Keys.Up:
                    SelectNextLink(false);
                    break;
                default:
                    return base.ProcessDialogKey(keyData);
            }
            return false;
        }
    }
}
